File: question207/C++/solution.cpp
class Solution {
	public:
		bool canFinish(int numCourses, vector<pair<int, int>>& prerequisites) {
			int inDegree[numCourses] = {0};	//记录每个节点的入度
			vector<int> graph[numCourses];
			for(int i = 0; i < prerequisites.size(); i++) {
				graph[prerequisites[i].second].push_back(prerequisites[i].first);
				inDegree[prerequisites[i].first]++;
			}
			int num = 0;	//记录加入拓扑序列的顶点数
			queue<int> q;
			for(int i = 0; i < numCourses; i++) {
				if(inDegree[i] == 0) {
					q.push(i);	//将所有入度为0的顶点入队
				}
			}
			while(!q.empty()) {
				int u = q.front();	//取队首顶点u
				q.pop();
				for(int i = 0; i < graph[u].size(); i++) {
					int v = graph[u][i];	//u的后继节点v
					inDegree[v]--;	//顶点v的入度减1
					if(inDegree[v] == 0) {
						q.push(v);	//顶点v的入度减为0则入队
					}
				}
//				graph[u].clear(); 
				num++;
			}
			if(num == numCourses){
				return true;	//加入拓扑序列的顶点数为numCourses，说明拓扑排序成功 
			}
			return false;	//加入拓扑序列的顶点数小于numCourses，说明拓扑排序失败 
		}
};

File: question207/Java/Solution.java
package question207.Java;

import java.util.LinkedList;
import java.util.Queue;

public class Solution {

    public boolean canFinish(int numCourses, int[][] prerequisites) {
        int[] inDegree = new int[numCourses];
        int[][] graph = new int[numCourses][numCourses];
        for(int i = 0; i < prerequisites.length; i++){
            graph[prerequisites[i][1]][prerequisites[i][0]] = 1;
            inDegree[prerequisites[i][0]]++;
        }
        Queue<Integer> queue = new LinkedList<>();
        for(int i = 0; i < numCourses; i++){
            if(inDegree[i] == 0){
                queue.add(i);
            }
        }
        int num = 0;
        while(!queue.isEmpty()){
            int u = queue.poll();
            for(int v = 0; v < numCourses; v++){
                if(graph[u][v] != 0){
                    inDegree[v]--;
                    if(inDegree[v] == 0){
                        queue.add(v);
                    }
                }
            }
            num++;
        }
        return num == numCourses;
    }
}
